### PR TITLE
Fix wcstoutf8()

### PR DESCRIPTION
**Pull request by krlmlr**
First pass at fixing wcstoutf8:  define n to be size of buffer, make sure buffer can't be overwritten

### CODE REVIEW

File: src/main/grep.c [hunk 1]
```diff
@@ -138,21 +138,19 @@ static SEXP mkCharWLen(const wchar_t *wc, int nc)
...
     wcsncpy(wt, wc, nc); wt[nc] = 0;
-    nb = wcstoutf8(NULL, wt, nc);
-    R_CheckStack2(sizeof(char)*(nb+1));
-    xi = (char *) alloca((nb+1)*sizeof(char));
-    wcstoutf8(xi, wt, nb + 1);
-    if (nb > INT_MAX)
-	error("R character strings are limited to 2^31-1 bytes");
+    nb = wcstoutf8(NULL, wt, INT_MAX);
+    R_CheckStack2(sizeof(char)*nb);
+    xi = (char *) alloca(nb*sizeof(char));
```
Comment (krlmlr) [inline src/main/grep.c]: This is inconsistent with `mkCharW()` which uses `Calloc()` (perhaps to reduce stress on the stack).

File: src/main/character.c [hunk 2]
```diff
@@ -1426,9 +1426,9 @@ SEXP attribute_hidden do_chartr(SEXP call, SEXP op, SEXP args, SEXP env)
 		    if (tbl) wc[j] = tbl->c_new;
 		}
 		if (ienc == CE_UTF8) {
-		    nb = (int) wcstoutf8(NULL, wc, 0);
+		    nb = (int) wcstoutf8(NULL, wc, INT_MAX);
```
Comment (krlmlr) [inline src/main/character.c]: Can we use a variable that is local to this block to avoid the cast?

File: src/main/character.c [hunk 3]
```diff
@@ -769,9 +769,9 @@ static SEXP wstripchars(const wchar_t * const inchar, int minlen, int usecl)
 		if (iswspace((int)wc[i])) mywcscpy(wc + i, wc + i + 1);
     }
 
-    int nb = (int) wcstoutf8(NULL, wc, 0);
-    char *cbuf = CallocCharBuf(nb+1);
-    wcstoutf8(cbuf, wc, nb + 1);
+    int nb = (int) wcstoutf8(NULL, wc, INT_MAX);
```
Comment (krlmlr) [inline src/main/character.c]: Can we declare as `size_t`?

**Comment by dmurdoch** [comment]
On 21/05/2017 4:09 PM, Kirill Müller wrote:
> *@krlmlr* commented on this pull request.
>
> Thanks. Looks like this PR will make the code much safer, I tried to err
> on the side of caution in my review.
>
> ------------------------------------------------------------------------
>
> In src/gnuwin32/extra.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117640703>:
>
>> @@ -447,15 +447,15 @@ SEXP do_normalizepath(SEXP call, SEXP op, SEXP args, SEXP rho)
>  					wtmp, &wtmp2)) && res <= 32768) {
>  		if ((res = GetLongPathNameW(wtmp, wlongpath, 32768))
>
> Do you want to use a meaningful name/expression for |32768|?

I don't know the source for that, so I'd rather not touch it now.
>
> ------------------------------------------------------------------------
>
> In src/main/sysutils.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117640724>:
>
>> @@ -400,7 +400,7 @@ SEXP attribute_hidden do_getenv(SEXP call, SEXP op, SEXP args, SEXP env)
>  		int n = wcslen(w), N = 3*n+1; /* UCS-2 maps to <=3 UTF-8 */
>
> Do we need |4*n + 1| here?

Yes, good catch.

>
> ------------------------------------------------------------------------
>
> In src/main/util.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117641007>:
>
>>  {
>      register size_t i, j;
> -    char buf[10], *b;
> -
> -    b = s ? s : buf;
> -    if(cvalue == 0) {*b = 0; return 0;}
> +    if (s) *s = 0;    /* Simplifies exit later */
>
> Can |n| be zero here?

I was assuming it never would be, but it's safer to check.

>
> ------------------------------------------------------------------------
>
> In src/main/util.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117641096>:
>
>>      char *t;
>      const wchar_t *p;
> -    if(s) {
> -	for(p = wc, t = s; ; p++) {
> -	    if (IS_SURROGATE_PAIR(*p,  *(p+1))) {
> -	    	Rwchar_t cvalue = ((*p & 0x3FF) << 10) + (*(p+1) & 0x3FF) + 0x010000;
> -	    	m = (ssize_t) Rwcrtomb32(t, cvalue);
> -	    	p++;
> -	    } else
> -	    	m  = (ssize_t) Rwcrtomb(t, *p);
> -	    if(m <= 0) break;
> -	    res += m;
> -	    if(res >= n) break;
> +    for(p = wc, t = s; ; p++) {
> +    	if (IS_SURROGATE_PAIR(*p, *(p+1))) {
> +    	    Rwchar_t cvalue =  ((*p & 0x3FF) << 10) + (*(p+1) & 0x3FF) + 0x010000;
> +	    m = Rwcrtomb32(t, cvalue, n - res);
>
> We could consider passing |n - res - 1| to make sure that there's always
> space for the next codepoint and the following null terminator (if
> |cvalue| is not zero).

I've changed the test in Rwcrtomb32 instead.

>
> ------------------------------------------------------------------------
>
> In src/main/grep.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117641126>:
>
>> @@ -1593,8 +1591,8 @@ char *pcre_string_adj(char *target, const char *orig, const char *repl,
>  			wc = (wchar_t *) alloca((nc+1)*sizeof(wchar_t));
>  			utf8towcs(wc, xi, nc + 1);
>  			for (j = 0; j < nc; j++) wc[j] = towctrans(wc[j], tr);
> -			nb = (int) wcstoutf8(NULL, wc, 0);
> -			wcstoutf8(xi, wc, nb + 1);
> +			nb = (int) wcstoutf8(NULL, wc, INT_MAX);
>
> Is there a guarantee that (new) |nb| is larger than (old) |nb| which is
> used to allocate |xi|? I'm having trouble understanding this code.

The original in R-devel was

     wcsncpy(wt, wc, nc); wt[nc] = 0;
     nb = wcstoutf8(NULL, wt, nc);
     R_CheckStack2(sizeof(char)*(nb+1));
     xi = (char *) alloca((nb+1)*sizeof(char));
     wcstoutf8(xi, wt, nb + 1);
     if (nb > INT_MAX)
	error("R character strings are limited to 2^31-1 bytes");

This has a couple of problems.  First, nc doesn't make sense in the 
first wcstoutf8 call.  It's the number of wide characters; what we want 
is the max size of buffer.  Second, in the new code, nb is the number of 
bytes to allocate including the terminating null.

The new code is

     wcsncpy(wt, wc, nc); wt[nc] = 0;
     nb = wcstoutf8(NULL, wt, INT_MAX);
     R_CheckStack2(sizeof(char)*nb);
     xi = (char *) alloca(nb*sizeof(char));
     wcstoutf8(xi, wt, nb);

We don't need the final test because nb is restricted to be less than 
INT_MAX.




>
> ------------------------------------------------------------------------
>
> In src/main/grep.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117641259>:
>
>> @@ -138,21 +138,19 @@ static SEXP mkCharWLen(const wchar_t *wc, int nc)
>      R_CheckStack2(sizeof(wchar_t)*(nc+1));
>      wt = (wchar_t *) alloca((nc+1)*sizeof(wchar_t));
>      wcsncpy(wt, wc, nc); wt[nc] = 0;
> -    nb = wcstoutf8(NULL, wt, nc);
> -    R_CheckStack2(sizeof(char)*(nb+1));
> -    xi = (char *) alloca((nb+1)*sizeof(char));
> -    wcstoutf8(xi, wt, nb + 1);
> -    if (nb > INT_MAX)
> -	error("R character strings are limited to 2^31-1 bytes");
> +    nb = wcstoutf8(NULL, wt, INT_MAX);
> +    R_CheckStack2(sizeof(char)*nb);
> +    xi = (char *) alloca(nb*sizeof(char));
>
> This is inconsistent with |mkCharW()| which uses |Calloc()| (perhaps to
> reduce stress on the stack).

Again, I won't change that in this patch.

>
> ------------------------------------------------------------------------
>
> In src/main/character.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117641300>:
>
>> @@ -1426,9 +1426,9 @@ SEXP attribute_hidden do_chartr(SEXP call, SEXP op, SEXP args, SEXP env)
>  		    if (tbl) wc[j] = tbl->c_new;
>  		}
>  		if (ienc == CE_UTF8) {
> -		    nb = (int) wcstoutf8(NULL, wc, 0);
> +		    nb = (int) wcstoutf8(NULL, wc, INT_MAX);
>
> Can we use a variable that is local to this block to avoid the cast?

Another case where I'd rather not change things.

>
> ------------------------------------------------------------------------
>
> In src/main/character.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117641376>:
>
>> @@ -769,9 +769,9 @@ static SEXP wstripchars(const wchar_t * const inchar, int minlen, int usecl)
>  		if (iswspace((int)wc[i])) mywcscpy(wc + i, wc + i + 1);
>      }
>
> -    int nb = (int) wcstoutf8(NULL, wc, 0);
> -    char *cbuf = CallocCharBuf(nb+1);
> -    wcstoutf8(cbuf, wc, nb + 1);
> +    int nb = (int) wcstoutf8(NULL, wc, INT_MAX);
>
> Can we declare as |size_t|?

Same as above.

Duncan

File: src/gnuwin32/extra.c [hunk 4]
```diff
@@ -447,15 +447,15 @@ SEXP do_normalizepath(SEXP call, SEXP op, SEXP args, SEXP rho)
 					wtmp, &wtmp2)) && res <= 32768) {
 		if ((res = GetLongPathNameW(wtmp, wlongpath, 32768))
```
Comment (krlmlr) [inline src/gnuwin32/extra.c]: #4

File: src/main/character.c (hunk 3 above)
Comment (krlmlr) [inline src/main/character.c]: #4

File: src/main/character.c (hunk 2 above)
Comment (krlmlr) [inline src/main/character.c]: #4

File: src/main/grep.c (hunk 1 above)
Comment (krlmlr) [inline src/main/grep.c]: #4

File: src/main/grep.c [hunk 5]
```diff
@@ -1593,8 +1591,8 @@ char *pcre_string_adj(char *target, const char *orig, const char *repl,
 			wc = (wchar_t *) alloca((nc+1)*sizeof(wchar_t));
 			utf8towcs(wc, xi, nc + 1);
 			for (j = 0; j < nc; j++) wc[j] = towctrans(wc[j], tr);
-			nb = (int) wcstoutf8(NULL, wc, 0);
-			wcstoutf8(xi, wc, nb + 1);
+			nb = (int) wcstoutf8(NULL, wc, INT_MAX);
```
Comment (krlmlr) [inline src/main/grep.c]: #4

File: src/main/util.c [hunk 6]
```diff
@@ -1324,64 +1323,54 @@ static const unsigned int utf8_table1[] =
...
-	    if(res >= n) break;
+    if (!n) return 0;
+    for(p = wc, t = s; ; p++) {
+    	if (IS_SURROGATE_PAIR(*p, *(p+1))) {
+    	    Rwchar_t cvalue =  ((*p & 0x3FF) << 10) + (*(p+1) & 0x3FF) + 0x010000;
+	    m = Rwcrtomb32(t, cvalue, n - res);
+	    p++;
+    	} else 
+    	    m = Rwcrtomb32(t, (Rwchar_t)(*p), n - res);
+    	if (!m) break;
```
Comment (krlmlr) [inline src/main/util.c]: What happens in this example:

```c
char target[1];
wcstoutf8(target, L"a", sizeof(target));
```

I'd expect `target` to contain only the null terminator, but desk-testing the current code seems to result in `target[0] == 'a'`.

Is there a place where we could add a unit test to this function, to test these corner cases? This would need to be an executable compiled against `libR` (?) and run as part of the test suite.

**Comment by dmurdoch** [comment]
On 21/05/2017 4:09 PM, Kirill Müller wrote:
> *@krlmlr* commented on this pull request.
>
> Thanks. Looks like this PR will make the code much safer, I tried to err
> on the side of caution in my review.

Thanks for all your comments.  I've now committed a final version to 
R-devel.

Duncan

**Comment by dmurdoch** [comment]
I did the commit in R-devel before seeing these...  more below.



On 22/05/2017 5:08 AM, Kirill Müller wrote:
> *@krlmlr* commented on this pull request.
>
> Thanks. I tend to integrate non-functional improvements with functional
> changes in my projects, but we can postpone them. I've opened issue #4
> <https://github.com/krlmlr/r-source/issues/4> to collect references to
> places where code can be improved, so that this can be done in a
> separate patch.
>
> ------------------------------------------------------------------------
>
> In src/gnuwin32/extra.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117695512>:
>
>> @@ -447,15 +447,15 @@ SEXP do_normalizepath(SEXP call, SEXP op, SEXP args, SEXP rho)
>  					wtmp, &wtmp2)) && res <= 32768) {
>  		if ((res = GetLongPathNameW(wtmp, wlongpath, 32768))
>
> #4 <https://github.com/krlmlr/r-source/issues/4>
>
> ------------------------------------------------------------------------
>
> In src/main/character.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117695627>:
>
>> @@ -769,9 +769,9 @@ static SEXP wstripchars(const wchar_t * const inchar, int minlen, int usecl)
>  		if (iswspace((int)wc[i])) mywcscpy(wc + i, wc + i + 1);
>      }
>
> -    int nb = (int) wcstoutf8(NULL, wc, 0);
> -    char *cbuf = CallocCharBuf(nb+1);
> -    wcstoutf8(cbuf, wc, nb + 1);
> +    int nb = (int) wcstoutf8(NULL, wc, INT_MAX);
>
> #4 <https://github.com/krlmlr/r-source/issues/4>
>
> ------------------------------------------------------------------------
>
> In src/main/character.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117695650>:
>
>> @@ -1426,9 +1426,9 @@ SEXP attribute_hidden do_chartr(SEXP call, SEXP op, SEXP args, SEXP env)
>  		    if (tbl) wc[j] = tbl->c_new;
>  		}
>  		if (ienc == CE_UTF8) {
> -		    nb = (int) wcstoutf8(NULL, wc, 0);
> +		    nb = (int) wcstoutf8(NULL, wc, INT_MAX);
>
> #4 <https://github.com/krlmlr/r-source/issues/4>
>
> ------------------------------------------------------------------------
>
> In src/main/grep.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117695685>:
>
>> @@ -138,21 +138,19 @@ static SEXP mkCharWLen(const wchar_t *wc, int nc)
>      R_CheckStack2(sizeof(wchar_t)*(nc+1));
>      wt = (wchar_t *) alloca((nc+1)*sizeof(wchar_t));
>      wcsncpy(wt, wc, nc); wt[nc] = 0;
> -    nb = wcstoutf8(NULL, wt, nc);
> -    R_CheckStack2(sizeof(char)*(nb+1));
> -    xi = (char *) alloca((nb+1)*sizeof(char));
> -    wcstoutf8(xi, wt, nb + 1);
> -    if (nb > INT_MAX)
> -	error("R character strings are limited to 2^31-1 bytes");
> +    nb = wcstoutf8(NULL, wt, INT_MAX);
> +    R_CheckStack2(sizeof(char)*nb);
> +    xi = (char *) alloca(nb*sizeof(char));
>
> #4 <https://github.com/krlmlr/r-source/issues/4>
>
> ------------------------------------------------------------------------
>
> In src/main/grep.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117695717>:
>
>> @@ -1593,8 +1591,8 @@ char *pcre_string_adj(char *target, const char *orig, const char *repl,
>  			wc = (wchar_t *) alloca((nc+1)*sizeof(wchar_t));
>  			utf8towcs(wc, xi, nc + 1);
>  			for (j = 0; j < nc; j++) wc[j] = towctrans(wc[j], tr);
> -			nb = (int) wcstoutf8(NULL, wc, 0);
> -			wcstoutf8(xi, wc, nb + 1);
> +			nb = (int) wcstoutf8(NULL, wc, INT_MAX);
>
> #4 <https://github.com/krlmlr/r-source/issues/4>
>
> ------------------------------------------------------------------------
>
> In src/main/util.c
> <https://github.com/krlmlr/r-source/pull/3#discussion_r117696314>:
>
>> -	    	m = (ssize_t) Rwcrtomb32(t, cvalue);
> -	    	p++;
> -	    } else
> -	    	m  = (ssize_t) Rwcrtomb(t, *p);
> -	    if(m <= 0) break;
> -	    res += m;
> -	    if(res >= n) break;
> +    if (!n) return 0;
> +    for(p = wc, t = s; ; p++) {
> +    	if (IS_SURROGATE_PAIR(*p, *(p+1))) {
> +    	    Rwchar_t cvalue =  ((*p & 0x3FF) << 10) + (*(p+1) & 0x3FF) + 0x010000;
> +	    m = Rwcrtomb32(t, cvalue, n - res);
> +	    p++;
> +    	} else
> +    	    m = Rwcrtomb32(t, (Rwchar_t)(*p), n - res);
> +    	if (!m) break;
>
> What happens in this example:
>
> char target[1];
> wcstoutf8(target, L"a", sizeof(target));
>
> I'd expect |target| to contain only the null terminator, but
> desk-testing the current code seems to result in |target[0] == 'a'|.

You're right, the test was still off by one.

>
> Is there a place where we could add a unit test to this function, to
> test these corner cases? This would need to be an executable compiled
> against |libR| (?) and run as part of the test suite.

We don't tend to have a good place for that.

Duncan

**Comment by krlmlr** [comment]
Thanks. Closing this PR, I have updated `master` of this fork so that you can simply branch off there.

I'll make myself familiar with the current testing infrastructure to see how unit tests could be integrated.
